File: src/everos_hermes/redaction.py
from __future__ import annotations

import copy
import json
import re
from typing import Any

CONTEXT_BLOCK_RE = re.compile(r"<(?P<tag>everos-context|memory-context)\b[^>]*>.*?</(?P=tag)>", re.IGNORECASE | re.DOTALL)
SENSITIVE_KEY_PATTERN = r"api[_-]?key|token|access[_-]?token|refresh[_-]?token|password|passwd|secret|authorization|credentials?|private[_-]?key"
SENSITIVE_KEY_RE = re.compile(rf"(^|[_\-\s.])({SENSITIVE_KEY_PATTERN})([_\-\s.]|$)", re.IGNORECASE)
SENSITIVE_ASSIGN_RE = re.compile(rf"(?P<prefix>[\"']?(?:{SENSITIVE_KEY_PATTERN})[\"']?\s*[:=]\s*)", re.IGNORECASE)
MAX_SANITIZED_ERROR_CHARS = 500
DIAGNOSTIC_FIELD_RE = re.compile(
    r"\s+(?:request[_-]?id|trace[_-]?id|span[_-]?id|status|code)=[^\s,;\}\]]+",
    re.IGNORECASE,
)
EMBEDDED_JSON_STRING_PATTERNS = [
    re.compile(r"(?P<prefix>[\"']?[A-Za-z_][\w.\-]*[\"']?\s*[:=]\s*\")(?P<value>(?:\\.|[^\"])*)\""),
    re.compile(r"(?P<prefix>[\"']?[A-Za-z_][\w.\-]*[\"']?\s*[:=]\s*')(?P<value>(?:\\.|[^'])*)'"),
]
SECRET_TEXT_PATTERNS = [
    re.compile(r"Authorization\s*:\s*Bearer\s+[^\s,;\]}]+", re.IGNORECASE),
    re.compile(r"\bBearer\s+[A-Za-z0-9._~+/=\-]+", re.IGNORECASE),
    re.compile(r"\bsk-[A-Za-z0-9._\-]{4,}\b"),
    re.compile(
        rf"(?P<prefix>[\"']?(?:{SENSITIVE_KEY_PATTERN})[\"']?\s*[:=]\s*\")(?P<value>(?:\\.|[^\"])*)(?P<suffix>\")",
        re.IGNORECASE,
    ),
    re.compile(
        rf"(?P<prefix>[\"']?(?:{SENSITIVE_KEY_PATTERN})[\"']?\s*[:=]\s*')(?P<value>(?:\\.|[^'])*)(?P<suffix>')",
        re.IGNORECASE,
    ),
    re.compile(
        rf"(?P<prefix>[\"']?(?:{SENSITIVE_KEY_PATTERN})[\"']?\s*[:=]\s*)(?P<value>[^\r\n]+)",
        re.IGNORECASE,
    ),
]


def is_sensitive_key(key: Any) -> bool:
    return bool(SENSITIVE_KEY_RE.search(str(key)))


def strip_context_blocks(text: str) -> str:
    return CONTEXT_BLOCK_RE.sub("", text)


def _split_diagnostic_suffix(value: str) -> tuple[str, str]:
    return value, "".join(match.group(0) for match in DIAGNOSTIC_FIELD_RE.finditer(value))


def _truncate_sanitized(text: str) -> str:
    if len(text) <= MAX_SANITIZED_ERROR_CHARS:
        return text
    return text[:MAX_SANITIZED_ERROR_CHARS] + "...[truncated]"


def _decode_escaped_string(value: str) -> str | None:
    try:
        decoded = json.loads(f'"{value}"')
    except Exception:
        return None
    return decoded if isinstance(decoded, str) else None


def _escape_string_value(value: str) -> str:
    encoded = json.dumps(value, ensure_ascii=False)
    return encoded[1:-1]


def _scrub_json_text(text: str) -> str | None:
    stripped = text.strip()
    if stripped[:1] not in {"{", "["}:
        return None
    try:
        parsed = json.loads(stripped)
    except Exception:
        return None
    return json.dumps(scrub_value(parsed), ensure_ascii=False, sort_keys=True, separators=(",", ":"))


def _redact_embedded_json_strings(text: str) -> str:
    redacted = text
    for pattern in EMBEDDED_JSON_STRING_PATTERNS:
        def replace(match: re.Match[str]) -> str:
            decoded = _decode_escaped_string(match.group("value"))
            if decoded is None:
                return match.group(0)
            scrubbed = _scrub_json_text(decoded)
            if scrubbed is None:
                return match.group(0)
            suffix = '"' if match.group(0).endswith('"') else "'"
            return f"{match.group('prefix')}{_escape_string_value(scrubbed)}{suffix}"

        redacted = pattern.sub(replace, redacted)
    return redacted


def _find_balanced_end(text: str, start: int) -> int | None:
    pairs = {"{": "}", "[": "]"}
    opener = text[start : start + 1]
    closer = pairs.get(opener)
    if closer is None:
        return None
    depth = 0
    quote: str | None = None
    escape = False
    for pos in range(start, len(text)):
        char = text[pos]
        if quote is not None:
            if escape:
                escape = False
            elif char == "\\":
                escape = True
            elif char == quote:
                quote = None
            continue
        if char in {"'", '"'}:
            quote = char
        elif char == opener:
            depth += 1
        elif char == closer:
            depth -= 1
            if depth == 0:
                return pos + 1
    return None


def _redact_sensitive_jsonish_assignments(text: str) -> str:
    out: list[str] = []
    cursor = 0
    for match in SENSITIVE_ASSIGN_RE.finditer(text):
        if match.start() < cursor:
            continue
        value_start = match.end()
        while value_start < len(text) and text[value_start].isspace():
            value_start += 1
        if value_start >= len(text) or text[value_start] not in {"{", "["}:
            continue
        value_end = _find_balanced_end(text, value_start)
        if value_end is None:
            continue
        out.append(text[cursor : match.start()])
        out.append(match.group("prefix"))
        out.append("[REDACTED]")
        cursor = value_end
    if not out:
        return text
    out.append(text[cursor:])
    return "".join(out)


def redact_text(text: str) -> str:
    raw = str(text or "")
    scrubbed = _scrub_json_text(raw)
    if scrubbed is not None:
        return scrubbed
    redacted = _redact_embedded_json_strings(raw)
    redacted = _redact_sensitive_jsonish_assignments(redacted)
    for pattern in SECRET_TEXT_PATTERNS:
        if "prefix" in pattern.groupindex:
            def replace_secret(match: re.Match[str]) -> str:
                suffix = match.groupdict().get("suffix")
                if suffix is not None:
                    return f"{match.group('prefix')}[REDACTED]{suffix}"
                _, diagnostic_suffix = _split_diagnostic_suffix(match.group("value"))
                return f"{match.group('prefix')}[REDACTED]{diagnostic_suffix}"

            redacted = pattern.sub(replace_secret, redacted)
        else:
            redacted = pattern.sub("[REDACTED]", redacted)
    return redacted


def scrub_value(value: Any) -> Any:
    if isinstance(value, str):
        return strip_context_blocks(redact_text(value))
    if isinstance(value, list):
        return [scrub_value(item) for item in value]
    if isinstance(value, tuple):
        return [scrub_value(item) for item in value]
    if isinstance(value, dict):
        out: dict[str, Any] = {}
        for key, val in value.items():
            key_text = str(key)
            out[key_text] = "[REDACTED]" if is_sensitive_key(key_text) else scrub_value(val)
        return out
    return copy.deepcopy(value)


def sanitized_error_message(exc: BaseException | str) -> str:
    return _truncate_sanitized(redact_text(str(exc)))


def error_payload(operation: str, exc: BaseException | str, *, retryable: bool = True) -> dict[str, Any]:
    return {
        "ok": False,
        "operation": operation,
        "status": "error",
        "error_code": exc.__class__.__name__ if isinstance(exc, BaseException) else "error",
        "message": sanitized_error_message(exc),
        "retryable": retryable,
        "suggested_next_actions": ["inspect EverOS status/search before retrying to avoid duplicate writes"],
    }

File: rust-version/src/redaction.rs
use regex::{Captures, Regex};
use serde_json::{Value, json};
use std::fmt::Display;
use std::sync::OnceLock;

const REDACTED: &str = "[REDACTED]";
const MAX_SANITIZED_ERROR_CHARS: usize = 500;
const SENSITIVE_KEY_PATTERN: &str = "api[_-]?key|token|access[_-]?token|refresh[_-]?token|password|passwd|secret|authorization|credentials?|private[_-]?key";

pub fn strip_context_blocks(text: &str) -> String {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| {
        Regex::new(r"(?is)<everos-context\b[^>]*>.*?</everos-context>|<memory-context\b[^>]*>.*?</memory-context>").unwrap()
    })
    .replace_all(text, "")
    .to_string()
}

pub fn redact_text(text: &str) -> String {
    static AUTH_HEADER_RE: OnceLock<Regex> = OnceLock::new();
    static BEARER_RE: OnceLock<Regex> = OnceLock::new();
    static SK_RE: OnceLock<Regex> = OnceLock::new();
    static KV_DOUBLE_RE: OnceLock<Regex> = OnceLock::new();
    static KV_SINGLE_RE: OnceLock<Regex> = OnceLock::new();
    static KV_UNQUOTED_RE: OnceLock<Regex> = OnceLock::new();

    let text = AUTH_HEADER_RE
        .get_or_init(|| Regex::new(r"(?i)Authorization\s*:\s*Bearer\s+[^\s,;\]}]+").unwrap())
        .replace_all(text, REDACTED)
        .to_string();
    let text = BEARER_RE
        .get_or_init(|| Regex::new(r"(?i)\bBearer\s+[A-Za-z0-9._~+/=\-]+").unwrap())
        .replace_all(&text, REDACTED)
        .to_string();
    let text = SK_RE
        .get_or_init(|| Regex::new(r"\bsk-[A-Za-z0-9._\-]{4,}\b").unwrap())
        .replace_all(&text, REDACTED)
        .to_string();
    let text = redact_embedded_json_strings(&text);
    let text = redact_sensitive_jsonish_assignments(&text);
    let text = KV_DOUBLE_RE
        .get_or_init(|| {
            Regex::new(r#"(?i)(?P<prefix>[\"']?(?:api[_-]?key|token|access[_-]?token|refresh[_-]?token|password|passwd|secret|authorization|credentials?|private[_-]?key)[\"']?\s*[:=]\s*\")(?P<value>(?:\\.|[^"])*)(?P<suffix>\")"#).unwrap()
        })
        .replace_all(&text, |caps: &Captures<'_>| redact_kv_match(caps))
        .to_string();
    let text = KV_SINGLE_RE
        .get_or_init(|| {
            Regex::new(r#"(?i)(?P<prefix>[\"']?(?:api[_-]?key|token|access[_-]?token|refresh[_-]?token|password|passwd|secret|authorization|credentials?|private[_-]?key)[\"']?\s*[:=]\s*')(?P<value>(?:\\.|[^'])*)(?P<suffix>')"#).unwrap()
        })
        .replace_all(&text, |caps: &Captures<'_>| redact_kv_match(caps))
        .to_string();
    KV_UNQUOTED_RE
        .get_or_init(|| {
            Regex::new(r#"(?i)(?P<prefix>[\"']?(?:api[_-]?key|token|access[_-]?token|refresh[_-]?token|password|passwd|secret|authorization|credentials?|private[_-]?key)[\"']?\s*[:=]\s*)(?P<value>[^\r\n]+)"#).unwrap()
        })
        .replace_all(&text, |caps: &Captures<'_>| redact_kv_match(caps))
        .to_string()
}

fn redact_embedded_json_strings(text: &str) -> String {
    static DOUBLE_RE: OnceLock<Regex> = OnceLock::new();
    static SINGLE_RE: OnceLock<Regex> = OnceLock::new();
    let text = DOUBLE_RE
        .get_or_init(|| {
            Regex::new("(?P<prefix>[\\\"']?[A-Za-z_][A-Za-z0-9_.-]*[\\\"']?\\s*[:=]\\s*\\\")(?P<value>(?:\\\\.|[^\\\"])*)\\\"").unwrap()
        })
        .replace_all(text, |caps: &Captures<'_>| redact_embedded_json_match(caps, "\""))
        .to_string();
    SINGLE_RE
        .get_or_init(|| {
            Regex::new("(?P<prefix>[\\\"']?[A-Za-z_][A-Za-z0-9_.-]*[\\\"']?\\s*[:=]\\s*')(?P<value>(?:\\\\.|[^'])*)'").unwrap()
        })
        .replace_all(&text, |caps: &Captures<'_>| redact_embedded_json_match(caps, "'"))
        .to_string()
}

fn redact_embedded_json_match(caps: &Captures<'_>, suffix: &str) -> String {
    let Some(full) = caps.get(0).map(|m| m.as_str()) else {
        return String::new();
    };
    let prefix = caps.name("prefix").map(|m| m.as_str()).unwrap_or_default();
    let value = caps.name("value").map(|m| m.as_str()).unwrap_or_default();
    let quoted = format!("\"{value}\"");
    let Ok(decoded) = serde_json::from_str::<String>(&quoted) else {
        return full.to_string();
    };
    let Some(scrubbed) = scrub_json_text(&decoded) else {
        return full.to_string();
    };
    let Ok(encoded) = serde_json::to_string(&scrubbed) else {
        return full.to_string();
    };
    let escaped = encoded
        .strip_prefix('"')
        .and_then(|inner| inner.strip_suffix('"'))
        .unwrap_or(encoded.as_str());
    format!("{prefix}{escaped}{suffix}")
}

fn redact_sensitive_jsonish_assignments(text: &str) -> String {
    static ASSIGN_RE: OnceLock<Regex> = OnceLock::new();
    let assign_re = ASSIGN_RE.get_or_init(|| {
        Regex::new(r#"(?i)(?P<prefix>[\"']?(?:api[_-]?key|token|access[_-]?token|refresh[_-]?token|password|passwd|secret|authorization|credentials?|private[_-]?key)[\"']?\s*[:=]\s*)"#).unwrap()
    });
    let mut out = String::new();
    let mut cursor = 0usize;
    let mut changed = false;
    for caps in assign_re.captures_iter(text) {
        let Some(full) = caps.get(0) else {
            continue;
        };
        if full.start() < cursor {
            continue;
        }
        let value_start = skip_whitespace(text, full.end());
        let Some(opener) = text[value_start..].chars().next() else {
            continue;
        };
        if opener != '{' && opener != '[' {
            continue;
        }
        let Some(value_end) = find_balanced_end(text, value_start) else {
            continue;
        };
        out.push_str(&text[cursor..full.start()]);
        out.push_str(
            caps.name("prefix")
                .map(|m| m.as_str())
                .unwrap_or(full.as_str()),
        );
        out.push_str(REDACTED);
        cursor = value_end;
        changed = true;
    }
    if !changed {
        return text.to_string();
    }
    out.push_str(&text[cursor..]);
    out
}

fn skip_whitespace(text: &str, mut index: usize) -> usize {
    while index < text.len() {
        let Some(ch) = text[index..].chars().next() else {
            return index;
        };
        if !ch.is_whitespace() {
            return index;
        }
        index += ch.len_utf8();
    }
    index
}

fn find_balanced_end(text: &str, start: usize) -> Option<usize> {
    let opener = text[start..].chars().next()?;
    let closer = match opener {
        '{' => '}',
        '[' => ']',
        _ => return None,
    };
    let mut depth = 0usize;
    let mut quote: Option<char> = None;
    let mut escape = false;
    for (offset, ch) in text[start..].char_indices() {
        if let Some(active_quote) = quote {
            if escape {
                escape = false;
            } else if ch == '\\' {
                escape = true;
            } else if ch == active_quote {
                quote = None;
            }
            continue;
        }
        if ch == '"' || ch == '\'' {
            quote = Some(ch);
        } else if ch == opener {
            depth += 1;
        } else if ch == closer {
            depth = depth.saturating_sub(1);
            if depth == 0 {
                return Some(start + offset + ch.len_utf8());
            }
        }
    }
    None
}

fn redact_kv_match(caps: &Captures<'_>) -> String {
    let prefix = caps.name("prefix").map(|m| m.as_str()).unwrap_or_default();
    let suffix = caps.name("suffix").map(|m| m.as_str()).unwrap_or_default();
    if !suffix.is_empty() {
        return format!("{prefix}{REDACTED}{suffix}");
    }
    let value = caps.name("value").map(|m| m.as_str()).unwrap_or_default();
    let diagnostic_suffix = diagnostic_suffix(value);
    format!("{prefix}{REDACTED}{diagnostic_suffix}")
}

fn diagnostic_suffix(value: &str) -> String {
    static DIAGNOSTIC_RE: OnceLock<Regex> = OnceLock::new();
    DIAGNOSTIC_RE
        .get_or_init(|| {
            Regex::new(
                r"(?i)\s+(?:request[_-]?id|trace[_-]?id|span[_-]?id|status|code)=[^\s,;\}\]]+",
            )
            .unwrap()
        })
        .find_iter(value)
        .map(|found| found.as_str())
        .collect()
}

pub fn sanitized_error_message(err: impl Display) -> String {
    truncate_sanitized(redact_text(&err.to_string()))
}

fn truncate_sanitized(text: String) -> String {
    if text.chars().count() <= MAX_SANITIZED_ERROR_CHARS {
        return text;
    }
    let mut out: String = text.chars().take(MAX_SANITIZED_ERROR_CHARS).collect();
    out.push_str("...[truncated]");
    out
}

pub fn error_payload(operation: &str, err: impl Display) -> Value {
    json!({
        "ok": false,
        "operation": operation,
        "status": "error",
        "error_code": "error",
        "message": sanitized_error_message(err),
        "retryable": true,
        "suggested_next_actions": ["inspect EverOS status/search before retrying to avoid duplicate writes"],
    })
}

pub fn scrub_value(value: &Value) -> Value {
    match value {
        Value::String(text) => Value::String(scrub_string(text)),
        Value::Array(items) => Value::Array(items.iter().map(scrub_value).collect()),
        Value::Object(map) => Value::Object(
            map.iter()
                .map(|(key, value)| {
                    let value = if is_sensitive_key(key) {
                        Value::String(REDACTED.to_string())
                    } else {
                        scrub_value(value)
                    };
                    (key.clone(), value)
                })
                .collect(),
        ),
        other => other.clone(),
    }
}

fn scrub_string(text: &str) -> String {
    let stripped = strip_context_blocks(text);
    if let Some(scrubbed) = scrub_json_text(&stripped) {
        return scrubbed;
    }
    redact_text(&stripped)
}

fn scrub_json_text(text: &str) -> Option<String> {
    let trimmed = text.trim();
    if !(trimmed.starts_with('{') || trimmed.starts_with('[')) {
        return None;
    }
    let parsed = serde_json::from_str::<Value>(trimmed).ok()?;
    serde_json::to_string(&scrub_value(&parsed)).ok()
}

fn is_sensitive_key(key: &str) -> bool {
    static KEY_RE: OnceLock<Regex> = OnceLock::new();
    KEY_RE
        .get_or_init(|| {
            Regex::new(&format!(
                r"(?i)(^|[_\-\s.])({SENSITIVE_KEY_PATTERN})([_\-\s.]|$)"
            ))
            .unwrap()
        })
        .is_match(key)
}
